index.js

```diff
@@ -5,52 +5,58 @@ const v = require('node-bindgen-loader')({
 
 const stringify = (msg) => JSON.stringify(msg, null, 2)
 
-const verifySignatures = (msgs, cb) => {
+const verifySignatures = (hmacKey, msgs, cb) => {
   if (!Array.isArray(msgs)) return "input must be an array of message objects";
```

Comment:
Oh, I didn't see this before, but we should definitely do

```js
if (!Array.isArray(msgs)) {
  cb(new Error("input must be an array of message objects.");
  return;
}
```

instead of

```js
if (!Array.isArray(msgs)) return "input must be an array of message objects";
```

Because the API should always give the output via `cb` (asynchronous), not via `return` (synchronous). Right now, if we would make a test of this function with `msgs = 3` (anything that is not an array), the `cb` would never be called, so it would seem like this function is forever stuck.

There is another issue here too, which is about the error format. We basically had 3 design choices:

- Throw an error (how ssb-validate worked)
- Return a string (what we were doing before the callback-rewrite)
- Pass an error on the 1st argument to the callback

What we're doing now is passing a string to the 1st argument to the callback (see line 14). What we should do is pass an error. So `cb(new Error("etc etc"))` not `cb("etc etc")`. The `new Error()` is a standard way of passing exceptions to the callback, and would give us stack traces too.

Reply:
Ah thanks for this explanation! 🖤 I will add a test for `msgs = 3` ;) and rewrite all the errors to ensure we're returning `new Error()` to the callback. Will have to take a look at how this is handled in `node_bindgen`.

Reply:
@staltz Looking more closely at this, would it be sufficient to continue returning the error as a string from the Rust function(s) and creating the `new(Error("string goes here"))` in `index.js` before passing it to the callback (in a similar way to your example above with `!Array`)? Or would that result in an inaccurate / incorrect stack trace in the error object?

Reply:
@mycognosist, yep! Rust can return a string and then the JS side just wraps it in an Error type. We don't need the stack traces from within Rust leaking out to JS (I don't even know if that's possible)